chips/stm32f4xx/src/spi.rs

```diff
@@ -0,0 +1,500 @@
+            CR1::DFF::CLEAR +
+            // Enable
+            CR1::SPE::SET,
+        );
+    }
+
+    fn is_busy(&self) -> bool {
+        self.registers.sr.is_set(SR::BSY)
+    }
+
+    fn write_byte(&self, out_byte: u8) {
+        // loop till TXE (Transmit Buffer Empty) becomes 1
+        while !self.registers.sr.is_set(SR::TXE) {}
+
+        self.registers.dr.modify(DR::DR.val(out_byte as u32));
+    }
+
+    fn read_byte(&self) -> u8 {
+        self.read_write_byte(0)
+    }
+
+    fn read_write_byte(&self, val: u8) -> u8 {
+        self.write_byte(val);
+
+        // loop till RXNE becomes 1
+        while !self.registers.sr.is_set(SR::RXNE) {}
+
+        self.registers.dr.read(DR::DR) as u8
+    }
+
+    fn read_write_bytes(
+        &self,
+        write_buffer: &'static mut [u8],
+        read_buffer: Option<&'static mut [u8]>,
+        len: usize,
+    ) -> ReturnCode {
+        // If busy, don't start
+        if self.is_busy() {
+            return ReturnCode::EBUSY;
+        }
+
+        self.read_write_bytes(Some(write_buffer), read_buffer, len)
+    }
+
+    // We *only* support 1Mhz
+    fn set_rate(&self, rate: u32) -> u32 {
+        if rate != 1_000_000 {
+            panic!("rate must be 1_000_000");
+        }
+
+        self.set_cr(|| {
+            // HSI is 16Mhz and Fpclk is also 16Mhz. 0b011 is Fpclk / 16
+            self.registers.cr1.modify(CR1::BR.val(0b011));
+        });
+
+        1_000_000
+    }
+
+    fn get_rate(&self) -> u32 {
+        if self.registers.cr1.read(CR1::BR) != 0b011 {
+            panic!("rate not set to 1_000_000");
+        }
+
+        1_000_000
+    }
+
+    fn set_clock(&self, polarity: ClockPolarity) {
+        self.set_polarity(polarity);
+    }
+
+    fn get_clock(&self) -> ClockPolarity {
+        self.get_polarity()
+    }
+
+    fn set_phase(&self, phase: ClockPhase) {
+        self.set_phase(phase);
+    }
+
+    fn get_phase(&self) -> ClockPhase {
+        self.get_phase()
+    }
+
+    fn hold_low(&self) {
+        self.active_slave.map(|p| {
+            p.get_pin().as_ref().map(|pin| {
+                pin.clear();
+            });
+        });
+    }
+
+    fn release_low(&self) {
+        self.active_slave.map(|p| {
+            p.get_pin().as_ref().map(|pin| {
+                pin.set();
+            });
+        });
+    }
+
+    fn specify_chip_select(&self, cs: Self::ChipSelect) {
+        self.set_active_slave(cs);
+    }
+}
+
+impl dma1::StreamClient for Spi<'a> {
+    fn transfer_done(&self, pid: dma1::Dma1Peripheral) {
+        if pid == self.tx_dma_pid {
+            self.disable_tx();
+        }
+
+        if pid == self.rx_dma_pid {
+            self.disable_rx();
+        }
+
+        self.transfers_in_progress
+            .set(self.transfers_in_progress.get() - 1);
+
+        if self.transfers_in_progress.get() == 0 {
+            self.release_low();
+
+            let tx_buffer = self.tx_dma.map_or(None, |tx_dma| tx_dma.return_buffer());
```

Comment:
needless `map_or`, should be `map` or `and_then`

Reply:
Switched to `and_then`. Thank you for catching this.